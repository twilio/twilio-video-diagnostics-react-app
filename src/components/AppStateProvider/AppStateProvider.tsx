import React, { createContext, useContext, useReducer, useEffect } from 'react';
import produce from 'immer';
import { VideoInputTest } from '@twilio/rtc-diagnostics';

export enum ActivePane {
  GetStarted,
  DeviceCheck,
  DeviceError,
  CameraTest,
  Connectivity,
  Quality,
  Results,
}

interface stateType {
  activePane: ActivePane;
  videoGranted: boolean;
  audioGranted: boolean;
  deviceError: null | Error;
  videoInputTestReport: null | VideoInputTest.Report;
}

type ACTIONTYPE =
  | { type: 'set-active-pane'; newActivePane: ActivePane }
  | { type: 'next-pane' }
  | { type: 'previous-pane' }
  | { type: 'set-devices'; devices: MediaDeviceInfo[] }
  | { type: 'set-device-error'; error: Error }
  | { type: 'set-video-test-report'; report: VideoInputTest.Report };

type AppStateContextType = {
  state: stateType;
  dispatch: React.Dispatch<ACTIONTYPE>;
};

export const initialState = {
  activePane: ActivePane.GetStarted,
  videoGranted: false,
  audioGranted: false,
  deviceError: null,
  videoInputTestReport: null,
};

export const AppStateContext = createContext<AppStateContextType>(null!);

export function useAppStateContext() {
  const context = useContext(AppStateContext);
  if (!context) {
    throw new Error('useAppStateContext must be used within an AppStateProvider');
  }
  return context;
}

/**
 * This reducer function helps with the logic used to decide the activePane,
 * as well as which panes are to be shown to the user, and when.
 */

export const appStateReducer = produce((draft: stateType, action: ACTIONTYPE) => {
  switch (action.type) {
    case 'set-active-pane':
      draft.activePane = action.newActivePane;
      break;

    case 'next-pane':
      switch (draft.activePane) {
        case ActivePane.GetStarted:
          if (draft.audioGranted && draft.videoGranted) {
            draft.activePane = ActivePane.CameraTest;
          } else {
            draft.activePane = ActivePane.DeviceCheck;
          }
          break;
        case ActivePane.DeviceCheck:
          if (draft.deviceError) {
            draft.activePane = ActivePane.DeviceError;
          } else {
            draft.activePane = ActivePane.CameraTest;
          }
          break;
        default:
          draft.activePane++;
          break;
      }
      break;

    case 'previous-pane':
      switch (draft.activePane) {
        case ActivePane.CameraTest:
          if (draft.audioGranted && draft.videoGranted) {
            draft.activePane = ActivePane.GetStarted;
          } else {
            draft.activePane = ActivePane.DeviceCheck;
          }
          break;
        default:
          draft.activePane--;
          break;
      }
      break;

    case 'set-devices':
      draft.audioGranted = action.devices.filter((d) => d.kind === 'audioinput').every((d) => d.label);
      draft.videoGranted = action.devices.filter((d) => d.kind === 'videoinput').every((d) => d.label);
      break;

    case 'set-device-error':
      draft.deviceError = action.error;
      draft.activePane = ActivePane.DeviceError;
      break;

    case 'set-video-test-report':
      draft.videoInputTestReport = action.report;
  }
});

export const AppStateProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(appStateReducer, initialState);

  useEffect(() => {
    navigator.mediaDevices.enumerateDevices().then((devices) => dispatch({ type: 'set-devices', devices }));
  }, []);

  return <AppStateContext.Provider value={{ state, dispatch }}>{children}</AppStateContext.Provider>;
};
