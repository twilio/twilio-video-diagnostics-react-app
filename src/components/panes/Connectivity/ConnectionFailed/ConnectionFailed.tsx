import { useAppStateContext } from '../../../AppStateProvider/AppStateProvider';
import { createStyles, makeStyles, Button, Container, Grid, Typography, Paper } from '@material-ui/core';
import { DownloadIcon } from '../../../../icons/DownloadIcon';
import { ErrorIcon } from '../../../../icons/ErrorIcon';
import { ViewIcon } from '../../../../icons/ViewIcon';

const useStyles = makeStyles((theme) =>
  createStyles({
    header: {
      float: 'left',
      [theme.breakpoints.down('md')]: {
        float: 'initial',
      },
    },
    heading: {
      position: 'relative',
      marginTop: '20px',
    },
    paperContainer: {
      float: 'right',
      marginRight: '1em',
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'center',
      [theme.breakpoints.down('md')]: {
        float: 'initial',
        justifyContent: 'center',
        margin: '0 0 2.5em 0',
      },
    },
    paper: {
      padding: '1.5em',
      borderRadius: '8px',
      maxWidth: '400px',
      [theme.breakpoints.down('md')]: {
        width: '100%',
      },
    },
    downloadButton: {
      clear: 'left',
      marginTop: '2em',
      '& svg': {
        position: 'relative',
        left: '-5px',
      },
      [theme.breakpoints.down('md')]: {
        clear: 'initial',
        marginBottom: '2em',
      },
    },
    viewButton: {
      marginTop: '2em',
      '& svg': {
        position: 'relative',
        left: '-5px',
      },
    },
  })
);

interface ConnectionFailedProps {
  openModal: () => void;
}

export function ConnectionFailed({ openModal }: ConnectionFailedProps) {
  const classes = useStyles();
  const { downloadFinalTestResults } = useAppStateContext();

  return (
    <>
      <Container>
        <div>
          <Grid item lg={5} className={classes.header}>
            <Typography variant="h1" gutterBottom className={classes.heading}>
              <ErrorIcon />
              Connection failed
            </Typography>

            <Typography variant="body1" gutterBottom>
              It's possible that you're behind a firewall that will need to be updated by your network administrator.
            </Typography>
          </Grid>

          <Grid item lg={5} className={classes.paperContainer}>
            <Paper className={classes.paper}>
              <Typography variant="body1" gutterBottom>
                To conduct a video call, your internet needs to be able to communicate with Twilio's cloud.
              </Typography>

              <Typography variant="body1">
                Learn{' '}
                <a href="https://www.twilio.com/docs/video/ip-addresses" target="_blank" rel="noopener noreferrer">
                  how to configure your firewall correctly
                </a>
                .
              </Typography>
            </Paper>
            <Button variant="outlined" onClick={openModal} className={classes.viewButton}>
              <ViewIcon />
              View detailed connection information
            </Button>
          </Grid>

          <Grid item lg={5}>
            <Typography variant="body1">
              Download report results to send to your network admin, update your internet, and re-run this test.
            </Typography>

            <Button
              variant="contained"
              color="primary"
              onClick={downloadFinalTestResults}
              className={classes.downloadButton}
            >
              <DownloadIcon />
              Download report results
            </Button>
          </Grid>
        </div>
      </Container>
    </>
  );
}
