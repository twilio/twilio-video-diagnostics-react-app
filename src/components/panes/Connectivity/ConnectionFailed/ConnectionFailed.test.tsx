import { Button } from '@material-ui/core';
import { shallow } from 'enzyme';
import { ConnectionFailed } from './ConnectionFailed';
import { useAppStateContext } from '../../../AppStateProvider/AppStateProvider';

jest.mock('../../../AppStateProvider/AppStateProvider');

const mockUseAppStateContext = useAppStateContext as jest.Mock<any>;
const mockDownloadFinalTestResults = jest.fn();

mockUseAppStateContext.mockImplementation(() => ({
  state: { activePane: 4, preflightTest: { report: 'mockReport' } },
  nextPane: jest.fn(),
  downloadFinalTestResults: mockDownloadFinalTestResults,
}));

const mockSetIsModalOpen = jest.fn();

describe('the ConnectionFailed component', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<ConnectionFailed openModal={mockSetIsModalOpen} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should download the test results file if "Download report results" is clicked on', () => {
    const wrapper = shallow(<ConnectionFailed openModal={mockSetIsModalOpen} />);

    wrapper.find(Button).at(0).simulate('click');

    expect(mockDownloadFinalTestResults).toHaveBeenCalled();
  });
});
