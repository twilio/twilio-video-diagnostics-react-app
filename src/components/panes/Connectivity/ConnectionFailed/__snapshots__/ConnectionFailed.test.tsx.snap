// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the ConnectionFailed component should render correctly 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Container))>
    <WithStyles(ForwardRef(Grid))
      alignItems="center"
      container={true}
      justifyContent="space-between"
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
        md={5}
      >
        <WithStyles(ForwardRef(Typography))
          gutterBottom={true}
          style={
            Object {
              "position": "relative",
            }
          }
          variant="h1"
        >
          <div
            className="makeStyles-errorIcon-3"
          >
            <ErrorIcon />
          </div>
          Connection failed
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(Typography))
          gutterBottom={true}
          variant="body1"
        >
          It's possible that you're behind a firewall that will need to be updated by your network administrator.
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(Typography))
          variant="body1"
        >
          Download report results to send to your network admin, update your internet, and re-run this test.
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(Button))
          className="makeStyles-downloadButton-4"
          color="primary"
          onClick={[Function]}
          variant="contained"
        >
          <DownloadIcon />
          Download report results
        </WithStyles(ForwardRef(Button))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        className="makeStyles-paperContainer-2"
        item={true}
        md={5}
      >
        <WithStyles(ForwardRef(Paper))
          className="makeStyles-paper-1"
        >
          <WithStyles(ForwardRef(Typography))
            gutterBottom={true}
            variant="body1"
          >
            To conduct a video call, your internet needs to be able to communicate with Twilio's cloud.
          </WithStyles(ForwardRef(Typography))>
          <WithStyles(ForwardRef(Typography))
            variant="body1"
          >
            Learn
             
            <a
              href="https://www.twilio.com/docs/video/ip-addresses"
              rel="noopener noreferrer"
              target="_blank"
            >
              how to configure your firewall correctly
            </a>
            .
          </WithStyles(ForwardRef(Typography))>
        </WithStyles(ForwardRef(Paper))>
        <WithStyles(ForwardRef(Button))
          className="makeStyles-viewButton-5"
          onClick={[MockFunction]}
          variant="outlined"
        >
          <ViewIcon />
          View detailed connection information
        </WithStyles(ForwardRef(Button))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Container))>
</Fragment>
`;
